code-build.js

```diff
@@ -50,25 +53,47 @@ async function waitForBuildEndTime(sdk, { id, logs }, nextToken) {
       cloudWatchLogs
         .getLogEvents({ logGroupName, logStreamName, startFromHead, nextToken })
         .promise()
-  ]);
-  // Pluck off the relevant state
-  const [current] = batch.builds;
-  const { nextForwardToken, events = [] } = cloudWatch;
-
-  // stdout the CloudWatchLog (everyone likes progress...)
-  // CloudWatchLogs have line endings.
-  // I trim and then log each line
-  // to ensure that the line ending is OS specific.
-  events.forEach(({ message }) => console.log(message.trimEnd()));
-
-  // We did it! We can stop looking!
-  if (current.endTime && !events.length) return current;
-
-  // More to do: Sleep for 5 seconds :)
-  await new Promise(resolve => setTimeout(resolve, wait));
+  ]).catch((err) => {
+    errorDetected = true;
+    errMessage = err;
+  });
 
-  // Try again
-  return waitForBuildEndTime(sdk, current, nextForwardToken);
+  if(errorDetected) {
+    //We caught an error in trying to make the AWS api call, and are now checking to see if it was just a rate limiting error
+    if(errMessage.message && errMessage.message.search('Rate exceeded') !== -1) {
+
+      //We were rate-limited, so add 15 seconds to the wait time
+      let newWait = wait + 15000;
+
+      //Sleep before trying again
+      await new Promise(resolve => setTimeout(resolve, newWait));
+
+      // Try again from the same token position
+      return waitForBuildEndTime({ ...sdk, wait: newWait }, { id, logs }, nextToken);
+    } else {
+      //The error returned from the API wasn't about rate limiting, so throw it as an actual error and fail the job
+      throw errMessage;
+    }
+  } else {
```

Comment:
Drop the else case :)
This is the purpose of the function
and that intention is best communicated by having it at the "top level".

(fat finger, wrong else case the first time)